marlowe-apps/proofspace/Network/ProofSpace.hs

```diff
@@ -0,0 +1,272 @@
+import Servant.Server.Internal.DelayedIO (DelayedIO, delayedFail, delayedFailFatal, withRequest)
+import Servant.Server.Internal.ErrorFormatter (MkContextWithErrorFormatter)
+import System.IO.Unsafe (unsafePerformIO)
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as LBS8
+
+-- | A request for a ProofSpace credential.
+data ProofRequest = ProofRequest
+  { reqProtocolVersion :: Integer
+  , reqPublicServiceDid :: String
+  , reqSubscriberConnectDid :: String
+  , reqActionId :: String
+  , reqActionInstanceId :: String
+  , reqActionEventId :: String
+  , reqActionParams :: Value
+  , reqReceivedCredentials :: [ReceivedCredential]
+  , reqIssuedProjects :: Value
+  , reqIssuedCredentials :: Value
+  }
+  deriving (Show)
+
+instance FromJSON ProofRequest where
+  parseJSON =
+    withObject "ProofRequest" $
+      \o ->
+        do
+          reqProtocolVersion <- o .: "protocolVersion"
+          reqPublicServiceDid <- o .: "publicServiceDid"
+          reqSubscriberConnectDid <- o .: "subscriberConnectDid"
+          reqActionId <- o .: "actionId"
+          reqActionInstanceId <- o .: "actionInstanceId"
+          reqActionEventId <- o .: "actionEventId"
+          reqActionParams <- o .: "actionParams"
+          reqReceivedCredentials <- o .: "receivedCredentials"
+          reqIssuedProjects <- o .: "issuedProjects"
+          reqIssuedCredentials <- o .: "issuedCredentials"
+          pure ProofRequest{..}
+
+-- | A received ProofSpace credential.
+data ReceivedCredential = ReceivedCredential
+  { schemaId :: String
+  , credentialId :: String
+  , credWalletId :: String
+  , fields :: Value
+  , utcIssuedAt :: Integer
+  }
+  deriving (Show)
+
+instance FromJSON ReceivedCredential where
+  parseJSON =
+    withObject "ReceivedCredential" $
+      \o ->
+        do
+          schemaId <- o .: "schemaId"
+          credentialId <- o .: "credentialId"
+          credWalletId <- o .: "credWalletId"
+          fields <- o .: "fields"
+          utcIssuedAt <- o .: "utcIssuedAt"
+          pure ReceivedCredential{..}
+
+-- | The response for a ProofSpace webhook.
+data ProofResponse = ProofResponse
+  { resServiceDid :: String
+  , resSubscriberConnectDid :: String
+  , resActionEventId :: String
+  , resIssuedCredentials :: [Value]
+  , resRevokedCredentials :: [Value]
+  , resOk :: Bool
+  }
+  deriving (Show)
+
+instance ToJSON ProofResponse where
+  toJSON ProofResponse{..} =
+    object
+      [ "serviceDid" .= resServiceDid
+      , "subscriberConnectDid" .= resSubscriberConnectDid
+      , "actionEventId" .= resActionEventId
+      , "issuedCredentials" .= resIssuedCredentials
+      , "revokedCredentials" .= resRevokedCredentials
+      , "ok" .= resOk
+      ]
+
+-- | Type for processing credentials.
+type ProofHandler = ProofRequest -> Handler ProofResponse
+
+-- | Type for the API.
+type ProofApi = "sync-issue" :> ProofRequestBody :> Post '[JSON] ProofResponse
```

Comment:
Is there an easy way to make the path `"sync-issue"` specifiable at runtime via the command-line for `main`?